### PR TITLE
[4.2] PHP8.2 Replace deprecated utf8 encode/decode to mb string

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for Issue #38258 and partially for #38922.

### Summary of Changes
This pr replaces the few calls to utf8_encode/utf8_decode to mbstring. It uses hardcoded the latin 1 character set (ISO-8859-1).

[Update]
It also makes the mbstring polyfill an explicit dependency, so we ensure that it still works on hosts, which do not have the mbstring module loaded.

~If we should support all, then 'ISO-8859-1' can be replaced with mb_list_encodings().~ See comment below why not!

### Testing Instructions
Login in the frontend, and go to the profile user page.

### Actual result BEFORE applying this Pull Request
A deprecated warning is shown.

### Expected result AFTER applying this Pull Request
No deprecated warning is shown.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by nikosdion** [comment]
>  If we should support all, then 'ISO-8859-1' can be replaced with mb_list_encodings().

No, we absolutely shouldn't! 

The reason `utf8_decode` was deprecated is that it only ever supported ISO-8859-1 as per https://wiki.php.net/rfc/remove_utf8_decode_and_utf8_encode in the Introduction at the very top of the page (emphasis mine):

> The built-in functions utf8_encode and utf8_decode convert strings **encoded in ISO-8859-1 (“Latin 1”)** to and from UTF-8, respectively. While this is sometimes a useful feature, they are commonly misunderstood

I am unwilling to consider this PR a good solution because `mbstring` is actually not a hard requirement for Joomla 4 per [the published technical requirements](https://downloads.joomla.org/technical-requirements). I know it's nonsensical and probably a mistake on that page.

Unless that page is amended to list mbstring as a hard requirement it makes far more sense to follow the approach in #38922 with a function which uses whatever is actually available — including a fallback to a pure PHP implementation. Do note that the code in #38922 is in fact following the “Throwing the kitchen sink at it” section of the [PHP RFC for `utf8_decode` deprecation](https://wiki.php.net/rfc/remove_utf8_decode_and_utf8_encode), i.e. the “official” way PHP core developers want us to polyfill this deprecated feature. It also uses `utf8_decode` on older PHP versions to remove surprises, so there's that.

As a second stage we should look at every instance of `utf8_decode` and decide if it _actually_ makes sense to be doing that (most of that code came from early PHP 5 versions which didn't support UTF8 natively) and whether we actually meant to use ISO-8859-1 (e.g. when handling URLs that's deliberate) instead of blindly trying to use whatever encoding. I believe that in most cases we'll just end up removing the decoding altogether.

**Comment by laoneo** [comment]
I really like the idea to recheck if the code still makes sense and then remove what is not really needed. For now to improve compatibility I would prefer to replace the deprecated functions with alternatives.

From a maintenance perspective I would not add two helper functions which do basically replace 5 native function calls, even when it requires to change the compatibility requirements.

**Comment by nikosdion** [comment]
@laoneo I was thinking of a one-two approach. First add a drop-in replacement marked as `@internal` and `@deprecated 5.0`, then go through the (few, to be fair) instances where we use that function, thus removing the need for the drop-in replacement and the drop-in replacement itself.

This would be perfectly viable with regards to semantic versioning as `@internal` clearly communicates “thou shalt not use” to 3PDs and `@deprecated` sets the expectation for dropping the darned thing on a very short horizon.

Why do that instead of just doing the second step? To make sure that _real world_ use of the replacement doesn't cause any issues in use cases we have never thought of or could have expected just by looking at the code. I am a bit conservative here but I have seen… _THINGS_ 🤣

**Comment by laoneo** [comment]
We have been discussing this in maintainers and in #38922. Making the mbstring polyfill an explicit dependency and converting the core code to mbstring is the least disruptive change to improve PHP 8.2 compatibility in the 4.x series.

File: components/com_users/src/Model/ProfileModel.php [hunk 1]
```diff
@@ -130,7 +130,8 @@ public function getForm($data = array(), $loadData = true)
         $username = $loadData ? $form->getValue('username') : $this->loadFormData()->username;
 
         if ($username) {
-            $isUsernameCompliant  = !(preg_match('#[<>"\'%;()&\\\\]|\\.\\./#', $username) || strlen(utf8_decode($username)) < 2
+            $isUsernameCompliant  = !(preg_match('#[<>"\'%;()&\\\\]|\\.\\./#', $username)
+                || strlen(mb_convert_encoding($username, 'ISO-8859-1', 'UTF-8')) < 2
```
Comment (HLeithner) [inline components/com_users/src/Model/ProfileModel.php:134]: We have several alternatives beside doing the conversion our self:
`StringHelper::strlen()`
`utf8_strlen()`
`mb_strlen()`

Comment (laoneo) [reply components/com_users/src/Model/ProfileModel.php:134]: StringHelper::strlen() and utf8_strlen() do look both that the use internally `utf8_decode`, so only mb_strlen would fit. Is my assumption right?

Comment (laoneo) [reply components/com_users/src/Model/ProfileModel.php:134]: I leave it as it is, moving to another function can be done later in another pr.

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on f3e4c206b23a1362217c1761c139dac265d6131c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39583">issues.joomla.org/tracker/joomla-cms/39583</a>.</sub>

**Comment by joomdonation** [comment]
I restored test result from @carlitorweb because it was lost for some reasons. RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39583">issues.joomla.org/tracker/joomla-cms/39583</a>.</sub>

**Comment by roland-d** [comment]
Thank you

**Pull request merged**